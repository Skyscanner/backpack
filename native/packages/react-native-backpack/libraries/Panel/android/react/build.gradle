apply plugin: 'com.android.library'
apply plugin: 'kotlin-android'

android {
  compileSdkVersion rootProject.ext.compileSdkVersion
  buildToolsVersion rootProject.ext.buildToolsVersion


  defaultConfig {
    minSdkVersion rootProject.ext.minSdkVersion
    targetSdkVersion rootProject.ext.targetSdkVersion
    versionCode 1
    versionName "1.0"
  }
}

dependencies{
  implementation "com.facebook.react:react-native:+" // From node_modules
  implementation project(':panel-native')
  implementation "org.jetbrains.kotlin:kotlin-stdlib-jdk7:$kotlin_version"
}
repositories {
  mavenCentral()
}
