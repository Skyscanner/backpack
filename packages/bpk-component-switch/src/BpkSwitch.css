@keyframes bpk-keyframe-spin{100%{transform:rotate(1turn)}}.bpk-switch{position:relative;display:inline-block}.bpk-switch__checkbox{position:absolute;opacity:0}.bpk-switch__checkbox:checked ~ .bpk-switch__switch--primary{background:#0770e3;background:var(--bpk-switch-checked-color, #0770e3)}.bpk-switch__checkbox:checked ~ .bpk-switch__switch--event{background:#5a489b}.bpk-switch__checkbox:checked ~ .bpk-switch__switch::before{left:1.25rem}.bpk-switch__label{margin-right:.375rem;line-height:1.875rem}.bpk-switch__switch{position:relative;float:right;display:block;width:3rem;height:1.875rem;transition:background 200ms linear;border-radius:.9375rem;background:#dddde5;cursor:pointer}.bpk-switch__switch::before{position:absolute;top:.125rem;left:.125rem;content:'';display:block;width:1.625rem;height:1.625rem;transition:left 200ms ease-out;border-radius:50%;background:#fff;box-shadow:0px 1px 3px 0px rgba(37,32,31,0.3)}
