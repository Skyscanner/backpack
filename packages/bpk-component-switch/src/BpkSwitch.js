/*
 * Backpack - Skyscanner's Design System
 *
 * Copyright 2016-2020 Skyscanner Ltd
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
/* @flow strict */

import PropTypes from 'prop-types';
import React, { type Node } from 'react';
import { cssModules } from 'bpk-react-utils';

import STYLES from './BpkSwitch.scss';

const getClassName = cssModules(STYLES);

export const SWITCH_TYPES = {
  primary: 'primary',
  event: 'event',
};

export type SwitchTypeValue = $Values<typeof SWITCH_TYPES>;

const switchTypeClassNames = {
  [SWITCH_TYPES.primary]: getClassName('bpk-switch__switch--primary'),
  [SWITCH_TYPES.event]: getClassName('bpk-switch__switch--event'),
};

export type Props = {
  label: Node,
  type: SwitchTypeValue,
  className: ?string,
};

const BpkSwitch = (props: Props) => {
  const { className, label, type, ...rest } = props;

  const switchClassNames = [
    getClassName('bpk-switch__switch'),
    switchTypeClassNames[type],
  ];

  return (
    <label className={getClassName('bpk-switch', className)}>
      {/* $FlowFixMe - inexact rest. See 'decisions/flowfixme.md'. */}
      <input
        type="checkbox"
        className={getClassName('bpk-switch__checkbox')}
        aria-label={label}
        {...rest}
      />
      <span aria-hidden className={getClassName('bpk-switch__label')}>
        {label}
      </span>
      <span aria-hidden className={switchClassNames.join(' ')} />
    </label>
  );
};

BpkSwitch.propTypes = {
  label: PropTypes.node.isRequired,
  type: PropTypes.oneOf([SWITCH_TYPES.primary, SWITCH_TYPES.event]),
  className: PropTypes.string,
};

BpkSwitch.defaultProps = {
  className: null,
  type: SWITCH_TYPES.primary,
};

export default BpkSwitch;
