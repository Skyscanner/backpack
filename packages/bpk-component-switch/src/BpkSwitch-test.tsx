/*
 * Backpack - Skyscanner's Design System
 *
 * Copyright 2016 Skyscanner Ltd
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { render, screen } from '@testing-library/react';

import BpkSwitch from './BpkSwitch';

describe('BpkSwitch', () => {
  it('should render correctly', () => {
    const { container } = render(<BpkSwitch ariaLabel="Switch" />);
    expect(screen.getByRole('checkbox')).toBeInTheDocument();
    expect(screen.getByRole('checkbox')).toHaveAttribute('aria-label', 'Switch');
    expect(container.querySelectorAll(".bpk-switch__switch")[0].className).toBe("bpk-switch__switch");
  });

  it('should render small switch', () => {
    const { container } = render(<BpkSwitch ariaLabel="Switch" small />);
    expect(screen.getByRole('checkbox')).toBeInTheDocument();
    expect(container.querySelectorAll(".bpk-switch__checkbox")[0].className).toBe("bpk-switch__checkbox");
    expect(container.querySelectorAll(".bpk-switch__switch--small")[0].className).toBe("bpk-switch__switch bpk-switch__switch--small");
  });

  it('should render correctly when checked', () => {
    const { container } = render(
      <BpkSwitch checked onChange={() => {}} ariaLabel="Switch" />,
    );
    expect(screen.getByRole('checkbox')).toBeInTheDocument();
    expect(container.querySelectorAll(".bpk-switch__checkbox")[0]).toHaveAttribute('checked');
  });

  it('should support custom class names', () => {
    const { container } = render(
      <BpkSwitch ariaLabel="Switch" className="custom-classname" />,
    );
    expect(screen.getByRole('checkbox')).toBeInTheDocument();
    expect(container.querySelectorAll(".custom-classname")[0]).toBeInTheDocument();
  });

  it('should support arbitrary props', () => {
    render(<BpkSwitch ariaLabel="Switch" data-testid="my-switch" />);
    expect(screen.getByRole('checkbox')).toBeInTheDocument();
    expect(screen.getAllByTestId('my-switch')).toHaveLength(1);
  });
});
