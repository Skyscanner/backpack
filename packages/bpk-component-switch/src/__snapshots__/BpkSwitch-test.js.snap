// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BpkSwitch should render correctly 1`] = `
<label
  className="bpk-switch"
>
  <input
    aria-label="Switch"
    className="bpk-switch__checkbox"
    type="checkbox"
  />
  <span
    aria-hidden={true}
    className="bpk-switch__label"
  >
    Switch
  </span>
  <span
    aria-hidden={true}
    className="bpk-switch__switch bpk-switch__switch--primary"
  />
</label>
`;

exports[`BpkSwitch should render correctly when checked 1`] = `
<label
  className="bpk-switch"
>
  <input
    aria-label="Switch"
    checked={true}
    className="bpk-switch__checkbox"
    type="checkbox"
  />
  <span
    aria-hidden={true}
    className="bpk-switch__label"
  >
    Switch
  </span>
  <span
    aria-hidden={true}
    className="bpk-switch__switch bpk-switch__switch--primary"
  />
</label>
`;

exports[`BpkSwitch should render event type 1`] = `
<label
  className="bpk-switch"
>
  <input
    aria-label="Switch"
    className="bpk-switch__checkbox"
    type="checkbox"
  />
  <span
    aria-hidden={true}
    className="bpk-switch__label"
  >
    Switch
  </span>
  <span
    aria-hidden={true}
    className="bpk-switch__switch bpk-switch__switch--event"
  />
</label>
`;

exports[`BpkSwitch should support arbitrary props 1`] = `
<label
  className="bpk-switch"
>
  <input
    aria-label="Switch"
    className="bpk-switch__checkbox"
    testID="123"
    type="checkbox"
  />
  <span
    aria-hidden={true}
    className="bpk-switch__label"
  >
    Switch
  </span>
  <span
    aria-hidden={true}
    className="bpk-switch__switch bpk-switch__switch--primary"
  />
</label>
`;

exports[`BpkSwitch should support custom class names 1`] = `
<label
  className="bpk-switch custom-classname"
>
  <input
    aria-label="Switch"
    className="bpk-switch__checkbox"
    type="checkbox"
  />
  <span
    aria-hidden={true}
    className="bpk-switch__label"
  >
    Switch
  </span>
  <span
    aria-hidden={true}
    className="bpk-switch__switch bpk-switch__switch--primary"
  />
</label>
`;
